core/src/main/java/org/elasticsearch/action/search/InitialSearchPhase.java

```diff
@@ -131,55 +139,94 @@ private void onShardFailure(final int shardIndex, @Nullable ShardRouting shard,
 
     @Override
     public final void run() throws IOException {
-        boolean success = shardExecutionIndex.compareAndSet(0, maxConcurrentShardRequests);
-        assert success;
-        for (int i = 0; i < maxConcurrentShardRequests; i++) {
-            SearchShardIterator shardRoutings = shardsIts.get(i);
-            if (shardRoutings.skip()) {
-                skipShard(shardRoutings);
-            } else {
-                performPhaseOnShard(i, shardRoutings, shardRoutings.nextOrNull());
+        for (final SearchShardIterator iterator : skippedShardsIts) {
+            if (iterator.skip()) {
```

Comment:
this can be an assertion, right? no need to check since we checked above?

Reply:
Yes, it’s leftover.